File: Company_wise_Mock/Microsoft/Coin Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        n = len(coins)
        
        amounts = [math.inf] * (amount + 1)
        amounts[0] = 0
        
        for i in range(1, amount + 1):
            for c in coins:
                remain_amount = i - c
                
                if remain_amount >= 0 and amounts[remain_amount] != math.inf:
                    amounts[i] = min(amounts[i], 1 + amounts[remain_amount])
        
        if amounts[amount] == math.inf:
            return -1
        
        return amounts[amount]

File: Daily Challenges/2021/March/Coin Change.cs
public class Solution {
    public int CoinChange(int[] coins, int amount) {
        int inf = 999999999;

        int[] dp = new int[amount + 1];
        Array.Fill(dp, inf);

        foreach (int c in coins)
        {
            if(c <= amount)
                dp[c] = 1;
        }

        dp[0] = 0;
        for(int i = 1; i <= amount; i++)
        {
            int mn = inf;
            foreach(int c in coins)
            {
                if (i >= c)
                    mn = Math.Min(mn, dp[i - c]);
            }

            if (mn < inf) dp[i] = mn + 1;
        }
        
        if(dp[amount] >= inf) return -1;
        
        return dp[amount];
    }
}
